setup.sh

```diff
@@ -1,45 +1,110 @@
+  exit 1
+fi
+
 # Ask about the generic CARDS version to use
 
-# Get the latest CARDS version from Nexus
-CARDS_VERSION=$(curl -s https://nexus.phenotips.org/nexus/content/repositories/releases/io/uhndata/cards/cards-modules/maven-metadata.xml | grep --max-count=1 '<release>' | cut '-d>' -f2 | cut '-d<' -f1)
-if [[ -z $CARDS_VERSION ]]
+# Get the available CARDS versions from Maven Central
+declare -a RELEASED_VERSIONS=( $(curl -s https://repo1.maven.org/maven2/io/uhndata/cards/cards-modules/maven-metadata.xml | grep '<version>' | tac | cut '-d>' -f2 | cut '-d<' -f1) )
+declare -i RELEASED_VERSIONS_COUNT=${#RELEASED_VERSIONS[*]}
+# Look on the local disk
+# First check that the Docker image exists on the local machine
+(docker image inspect cards/cards:latest >/dev/null 2>/dev/null) && (LOCAL_CARDS_VERSION=$(docker run --rm  --entrypoint /bin/sh cards/cards:latest -c "ls /root/.m2/repository/io/uhndata/cards/cards-parent/ 2>/dev/null | grep SNAPSHOT"))
+declare VERSION
+declare HAS_DEFAULT
+for (( i=0 ; i<RELEASED_VERSIONS_COUNT; i++ ))
+do
+  VERSION="${VERSION} ${RELEASED_VERSIONS[$i]} ${RELEASED_VERSIONS[$i]} ${HAS_DEFAULT:-ON} "
+  HAS_DEFAULT=OFF
+done
+if [[ -n $LOCAL_CARDS_VERSION ]]
+then
+  VERSION="${LOCAL_CARDS_VERSION} ${LOCAL_CARDS_VERSION} ${HAS_DEFAULT:-ON} ${VERSION}"
+  HAS_DEFAULT=OFF
+fi
+VERSION="${VERSION} other other ${HAS_DEFAULT:-ON}"
+CARDS_VERSION=$(whiptail --backtitle "New CARDS repository setup" --title "Base CARDS version" --radiolist --notags "Which version of the CARDS platform should be used?" 38 78 30 $VERSION 3>&1 1>&2 2>&3)
+if [[ $? == 1 ]]
 then
-  # If Nexus isn't available, try to look on the local disk
-  CARDS_VERSION=$(grep --max-count=1 '^  <version>' ../cards/pom.xml | cut '-d>' -f2 | cut '-d<' -f1)
+  exit 1
+fi
+if [[ $CARDS_VERSION == "other" ]]
+then
+  until [[ $CARDS_VERSION != "other" && -n $CARDS_VERSION ]]
+  do
+    CARDS_VERSION=$(whiptail --backtitle "New CARDS repository setup" --title "Base CARDS version" --inputbox "Which version of the CARDS platform should be used? e.g. 0.9.22" 8 78 3>&1 1>&2 2>&3)
+    if [[ $? == 1 ]]
+    then
+      exit 1
+    fi
+  done
 fi
-echo
-echo "Which version of the CARDS platform should be used?"
-read -e -i "$CARDS_VERSION" CARDS_VERSION
 
 # Ask about the project name
 
-echo "What is the project's codename, e.g. cards4sparc, cards4heracles"
-read -e -i cards4 PROJECT_CODENAME
+declare PROJECT_CODENAME
+while [[ -z $PROJECT_CODENAME ]]
+do
+  PROJECT_CODENAME=$(whiptail --backtitle "New CARDS repository setup" --title "Project name" --inputbox "What is the project's codename? No uppercase letters are allowed, e.g. cards4sparc, cards4lfs" 8 78 cards4 3>&1 1>&2 2>&3)
+  if [[ $? == 1 ]]
+  then
+    exit 1
+  elif [[ ${PROJECT_CODENAME} != ${PROJECT_CODENAME,,} ]]
+  then
+    PROJECT_CODENAME=
+  fi
+done
 
 # Remove the cards4 prefix to get the short unprefixed name
 PROJECT_SHORTNAME=${PROJECT_CODENAME#cards4}
+if [[ $? == 1 ]]
+then
+  exit 1
+fi
```

Comment:
What is the purpose of this?

Reply:
That's the status code for choosing "Cancel". I consider that pressing Cancel in a dialog shout terminate the script. A pending improvement for the next PR is to gather all the data first, then perform all the chances, so that pushing Cancel at a later stage doesn't leave the repo in an intermediary state.

Reply:
Right, but isn't that already handled by lines 76-78

```bash
while [[ -z $PROJECT_CODENAME ]]
do
  PROJECT_CODENAME=$(whiptail --backtitle "New CARDS repository setup" --title "Project name" --inputbox "What is the project's codename? No uppercase letters are allowed, e.g. cards4sparc, cards4lfs" 8 78 cards4 3>&1 1>&2 2>&3)
  if [[ $? == 1 ]]
  then
    exit 1
  elif [[ ${PROJECT_CODENAME} != ${PROJECT_CODENAME,,} ]]
  then
    PROJECT_CODENAME=
  fi
done
```

Reply:
You're right, it is.